Error: converting tfplan to CAI assets: .. Error resolving image name ''"

**What do I see?**
When running a policy validate on our terraform, we get the following error.

`Error: converting tfplan to CAI assets: converting resource google_compute_disk: converting asset: Error resolving image name '': Error checking if image exists: json: invalid use of ,string struct tag, trying to unmarshal "projects/acmck-02-sandbox/global/images/" into uint64`

**What do I expect**
The policy validation  to output a violation(s) or report "No Violations".

**Reproduce**
Attach a disk to GCE instance with no image defined

```
resource "google_compute_disk" "bastion-disk" {
  name = "${var.name}-disk"
  type = "pd-ssd"
  zone = "us-central1-a"
  size = 100
}
```

and attach inside the instance resource:

```
    attached_disk {
      source      = google_compute_disk.bastion-disk.self_link
  }
```

**versions**

```
bash-5.0# terraform version
Terraform v0.12.16
+ provider.google v2.17.0
+ provider.google-beta v2.20.0
+ provider.kubernetes v1.10.0
+ provider.null v2.1.2
+ provider.random v2.2.1

Your version of Terraform is out of date! The latest version
is 0.12.17. You can update by downloading from https://www.terraform.io/downloads.html

bash-5.0# terraform-validator version
Build version: 2019-11-07
```
Happens on alpine or debain base image based containers.

```
bash-5.0# uname -a
Linux 9a1661522e53 4.9.184-linuxkit #1 SMP Tue Jul 2 22:58:16 UTC 2019 x86_64 Linux
```

Moving contents of ____ here:

When I merge https://github.com/GoogleCloudPlatform/terraform-validator/pull/119 into `terraform-validator`'s `main` branch and run its tests, I receive the following stack trace:

```
--- FAIL: TestReadPlannedAssetsCoverage (0.02s)
    --- FAIL: TestReadPlannedAssetsCoverage/example_compute_disk_empty (0.02s)
panic: nil context [recovered]
	panic: nil context

goroutine 57 [running]:
testing.tRunner.func1.2(0x403f400, 0x4d93e90)
	/usr/local/Cellar/go/1.16.6/libexec/src/testing/testing.go:1143 +0x332
testing.tRunner.func1(0xc000288f00)
	/usr/local/Cellar/go/1.16.6/libexec/src/testing/testing.go:1146 +0x4b6
panic(0x403f400, 0x4d93e90)
	/usr/local/Cellar/go/1.16.6/libexec/src/runtime/panic.go:965 +0x1b9
net/http.(*Request).WithContext(...)
	/usr/local/Cellar/go/1.16.6/libexec/src/net/http/request.go:356
golang.org/x/net/context/ctxhttp.Do(0x0, 0x0, 0x636b720, 0xc000140400, 0x406b560, 0x1, 0xc0021a8580)
	/Users/berlinsky/src/go/pkg/mod/golang.org/x/net@v0.0.0-20210503060351-7fd8e65b6420/context/ctxhttp/ctxhttp.go:27 +0x34f
golang.org/x/oauth2/internal.doTokenRoundTrip(0x0, 0x0, 0xc000140400, 0x48, 0xc0001dbdd0, 0x18)
	/Users/berlinsky/src/go/pkg/mod/golang.org/x/oauth2@v0.0.0-20210427180440-81ed05c6b58c/internal/token.go:232 +0x7b
golang.org/x/oauth2/internal.RetrieveToken(0x0, 0x0, 0xc0007227d0, 0x48, 0xc0001dbdd0, 0x18, 0x484c1d7, 0x23, 0xc002110690, 0x1, ...)
	/Users/berlinsky/src/go/pkg/mod/golang.org/x/oauth2@v0.0.0-20210427180440-81ed05c6b58c/internal/token.go:202 +0xe9
golang.org/x/oauth2.retrieveToken(0x0, 0x0, 0xc000310690, 0xc002110690, 0xc002110760, 0x403f400, 0xc002110698)
	/Users/berlinsky/src/go/pkg/mod/golang.org/x/oauth2@v0.0.0-20210427180440-81ed05c6b58c/token.go:157 +0x92
golang.org/x/oauth2.(*tokenRefresher).Token(0xc001e9fad0, 0x100f93b, 0x8, 0x203000)
	/Users/berlinsky/src/go/pkg/mod/golang.org/x/oauth2@v0.0.0-20210427180440-81ed05c6b58c/oauth2.go:270 +0x1e5
golang.org/x/oauth2.(*reuseTokenSource).Token(0xc000dfb6c0, 0x0, 0x0, 0x0)
	/Users/berlinsky/src/go/pkg/mod/golang.org/x/oauth2@v0.0.0-20210427180440-81ed05c6b58c/oauth2.go:304 +0xaf
golang.org/x/oauth2.(*Transport).RoundTrip(0xc000dfb700, 0xc000140100, 0x0, 0x0, 0x0)
	/Users/berlinsky/src/go/pkg/mod/golang.org/x/oauth2@v0.0.0-20210427180440-81ed05c6b58c/transport.go:45 +0x9e
net/http.send(0xc000140100, 0x4de4b60, 0xc000dfb700, 0x0, 0x0, 0x0, 0xc0006e2f60, 0xc002110be8, 0x1, 0x0)
	/usr/local/Cellar/go/1.16.6/libexec/src/net/http/client.go:251 +0x454
net/http.(*Client).send(0xc001e9fb60, 0xc000140100, 0x0, 0x0, 0x0, 0xc0006e2f60, 0x0, 0x1, 0x0)
	/usr/local/Cellar/go/1.16.6/libexec/src/net/http/client.go:175 +0xff
net/http.(*Client).do(0xc001e9fb60, 0xc000140100, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.16.6/libexec/src/net/http/client.go:717 +0x45f
net/http.(*Client).Do(...)
	/usr/local/Cellar/go/1.16.6/libexec/src/net/http/client.go:585
google.golang.org/api/internal/gensupport.SendRequest(0x0, 0x0, 0xc001e9fb60, 0xc000140100, 0xc002110f78, 0x55, 0x0)
	/Users/berlinsky/src/go/pkg/mod/google.golang.org/api@v0.46.0/internal/gensupport/send.go:26 +0xd0
google.golang.org/api/compute/v1.(*ImagesGetFromFamilyCall).doRequest(0xc0021110e8, 0x47ebfa7, 0x4, 0x0, 0x1010098, 0x30)
	/Users/berlinsky/src/go/pkg/mod/google.golang.org/api@v0.46.0/compute/v1/compute-gen.go:74226 +0x809
google.golang.org/api/compute/v1.(*ImagesGetFromFamilyCall).Do(0xc0021110e8, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/berlinsky/src/go/pkg/mod/google.golang.org/api@v0.46.0/compute/v1/compute-gen.go:74238 +0xa5
github.com/GoogleCloudPlatform/terraform-google-conversion/google.resolveImageFamilyExists(0xc00030f200, 0x47ee6c1, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/berlinsky/src/clients/google/usb/terraform-google-conversion/google/image.go:63 +0x10a
github.com/GoogleCloudPlatform/terraform-google-conversion/google.resolveImage(0xc00030f200, 0x47ee6c1, 0x6, 0x0, 0x0, 0x0, 0x0, 0x9, 0xc0007f0d00, 0x34, ...)
	/Users/berlinsky/src/clients/google/usb/terraform-google-conversion/google/image.go:189 +0x17d2
github.com/GoogleCloudPlatform/terraform-google-conversion/google.resourceComputeDiskEncoder(0x4e81758, 0xc002031ad0, 0x47a3780, 0xc00030f200, 0xc002031b90, 0x1, 0x0, 0x0)
	/Users/berlinsky/src/clients/google/usb/terraform-google-conversion/google/compute_disk.go:373 +0x5b4
```

It appears that we reach [google/image.go line 48](https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/master/google/image.go#L48), which attempts to make a request containing an empty string `name`, which leads to a `panic`.

I'm not familiar with how modular-magician generates this file, so I'm afraid my debugging might cease to be useful beyond this point.

This appears to be an error in the implementation of [FakeResourceData.GetOk](https://github.com/GoogleCloudPlatform/terraform-validator/blob/f299671d7ef7042d28ac0bd863169b40a885ec2c/converters/google/fake_resource_data.go#L54). When I add the following test case, I receive test failures, which are not expected:

```
func TestFakeResourceData_getOkNotOkStrVal(t *testing.T) {
	p := provider.Provider()

	values := map[string]interface{}{
		"name":  "test-disk",
		"type":  "pd-ssd",
		"zone":  "us-central1-a",
		"physical_block_size_bytes": 4096,
	}
	d := NewFakeResourceData(
		"google_compute_disk",
		p.ResourcesMap["google_compute_disk"].Schema,
		values,
	)
	res, ok := d.GetOk("image")
	assert.Nil(t, res)
	assert.False(t, ok)
}
```

```
--- FAIL: TestFakeResourceData_getOkNotOkStrVal (0.00s)
    fake_resource_data_test.go:136:
        	Error Trace:	fake_resource_data_test.go:136
        	Error:      	Expected nil, but got: ""
        	Test:       	TestFakeResourceData_getOkNotOkStrVal
```

--- a/converters/google/fake_resource_data.go
+++ b/converters/google/fake_resource_data.go
@@ -19,12 +19,16 @@ package google
 
 import (
 	"fmt"
+	"reflect"
 	"strconv"
 	"strings"
 	"time"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
+
+// Must be set to the same value as the internal typeObject const
+const typeObject schema.ValueType = 8
 
 // Compare to https://github.com/hashicorp/terraform-plugin-sdk/blob/97b4465/helper/schema/resource_data.go#L15
 type FakeResourceData struct {
@@ -52,7 +56,7 @@ func (d *FakeResourceData) Get(name string) interface{} {
 // Get reads a single field by key and returns a boolean indicating
 // whether the field exists.
 func (d *FakeResourceData) GetOk(name string) (interface{}, bool) {
-	res, err := d.reader.ReadField(strings.Split(name, "."))
+	r, err := d.reader.ReadField(strings.Split(name, "."))
 	if err != nil {
 		return nil, false
 	}
@@ -63,7 +67,26 @@ func (d *FakeResourceData) GetOk(name string) (interface{}, bool) {
 		return nil, false
 	}
 
-	return res.ValueOrZero(schemaPath[len(schemaPath)-1]), res.Exists && !res.Computed
+	exists := r.Exists && !r.Computed
+	s := schemaPath[len(schemaPath)-1]
+	value := r.Value
+
+	if value == nil && s != nil {
+		value = r.ValueOrZero(s)
+	}
+
+	if exists {
+		// Verify that it's not the zero-value
+		zero := s.Type.Zero()
+
+		if eq, ok := value.(schema.Equal); ok {
+			exists = !eq.Equal(zero)
+		} else {
+			exists = !reflect.DeepEqual(value, zero)
+		}
+	}
+
+	return value, exists
 }
 
 // GetOkExists currently just calls GetOk but should be updated to support
@@ -101,8 +124,6 @@ func NewFakeResourceData(kind string, resourceSchema map[string]*schema.Schema, 
 // NOTE: This function was copied from the terraform library:
 // github.com/hashicorp/terraform/helper/schema/field_reader.go
 func addrToSchema(addr []string, schemaMap map[string]*schema.Schema) []*schema.Schema {
-	const typeObject = 999
-
 	current := &schema.Schema{
 		Type: typeObject,
 		Elem: schemaMap,

--- a/converters/google/fake_resource_data.go
+++ b/converters/google/fake_resource_data.go
@@ -30,6 +30,17 @@ import (
 // Must be set to the same value as the internal typeObject const
 const typeObject schema.ValueType = 8
 
+
+// This is more or less equivalent to the internal getResult struct
+// used by schema.ResourceData
+type getResult struct {
+	Value interface{}
+	Computed bool
+	Exists bool
+	Schema *schema.Schema
+}
+
+
 // Compare to https://github.com/hashicorp/terraform-plugin-sdk/blob/97b4465/helper/schema/resource_data.go#L15
 type FakeResourceData struct {
 	reader schema.FieldReader
@@ -47,6 +58,36 @@ func (d *FakeResourceData) Id() string {
 	return ""
 }
 
+func (d *FakeResourceData) getRaw(key string) getResult {
+	var parts []string
+	if key != "" {
+		parts = strings.Split(key, ".")
+	}
+	return d.get(parts)
+}
+
+func (d *FakeResourceData) get(addr []string) getResult {
+	r, err := d.reader.ReadField(addr)
+	if err != nil {
+		panic(err)
+	}
+
+	var s *schema.Schema
+	if schemaPath := addrToSchema(addr, d.schema); len(schemaPath) > 0 {
+		s = schemaPath[len(schemaPath)-1]
+	}
+	if r.Value == nil && s != nil {
+		r.Value = r.ValueOrZero(s)
+	}
+
+	return getResult{
+		Value: r.Value,
+		Computed: r.Computed,
+		Exists: r.Exists,
+		Schema: s,
+	}
+}
+
 // Get reads a single field by key.
 func (d *FakeResourceData) Get(name string) interface{} {
 	val, _ := d.GetOk(name)
@@ -56,28 +97,13 @@ func (d *FakeResourceData) Get(name string) interface{} {
 // Get reads a single field by key and returns a boolean indicating
 // whether the field exists.
 func (d *FakeResourceData) GetOk(name string) (interface{}, bool) {
-	r, err := d.reader.ReadField(strings.Split(name, "."))
-	if err != nil {
-		return nil, false
-	}
-
-	addr := strings.Split(name, ".")
-	schemaPath := addrToSchema(addr, d.schema)
-	if len(schemaPath) == 0 {
-		return nil, false
-	}
-
+	r := d.getRaw(name)
 	exists := r.Exists && !r.Computed
-	s := schemaPath[len(schemaPath)-1]
-	value := r.Value
-
-	if value == nil && s != nil {
-		value = r.ValueOrZero(s)
-	}
 
 	if exists {
 		// Verify that it's not the zero-value
-		zero := s.Type.Zero()
+		value := r.Value
+		zero := r.Schema.Type.Zero()
 
 		if eq, ok := value.(schema.Equal); ok {
 			exists = !eq.Equal(zero)
@@ -86,13 +112,13 @@ func (d *FakeResourceData) GetOk(name string) (interface{}, bool) {
 		}
 	}
 
-	return value, exists
+	return r.Value, exists
 }
 
-// GetOkExists currently just calls GetOk but should be updated to support
-// schema.ResourceData.GetOkExists functionality.
-func (d *FakeResourceData) GetOkExists(name string) (interface{}, bool) {
-	return d.GetOk(name)
+func (d *FakeResourceData) GetOkExists(key string) (interface{}, bool) {
+	r := d.getRaw(key)
+	exists := r.Exists && !r.Computed
+	return r.Value, exists
 }
 
 // These methods are required by some mappers but we don't actually have (or need)

--- a/converters/google/fake_resource_data_test.go
+++ b/converters/google/fake_resource_data_test.go
@@ -24,9 +24,9 @@ func TestFakeResourceData_kind(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name": "test-disk",
-		"type": "pd-ssd",
-		"zone": "us-central1-a",
+		"name":  "test-disk",
+		"type":  "pd-ssd",
+		"zone":  "us-central1-a",
 		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
@@ -42,9 +42,9 @@ func TestFakeResourceData_id(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name": "test-disk",
-		"type": "pd-ssd",
-		"zone": "us-central1-a",
+		"name":  "test-disk",
+		"type":  "pd-ssd",
+		"zone":  "us-central1-a",
 		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
@@ -60,9 +60,9 @@ func TestFakeResourceData_get(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name": "test-disk",
-		"type": "pd-ssd",
-		"zone": "us-central1-a",
+		"name":  "test-disk",
+		"type":  "pd-ssd",
+		"zone":  "us-central1-a",
 		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
@@ -78,9 +78,9 @@ func TestFakeResourceData_getOkOk(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name": "test-disk",
-		"type": "pd-ssd",
-		"zone": "us-central1-a",
+		"name":  "test-disk",
+		"type":  "pd-ssd",
+		"zone":  "us-central1-a",
 		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
@@ -98,9 +98,9 @@ func TestFakeResourceData_getOkNonexistentField(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name": "test-disk",
-		"type": "pd-ssd",
-		"zone": "us-central1-a",
+		"name":  "test-disk",
+		"type":  "pd-ssd",
+		"zone":  "us-central1-a",
 		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}

--- a/converters/google/fake_resource_data_test.go
+++ b/converters/google/fake_resource_data_test.go
@@ -24,10 +24,10 @@ func TestFakeResourceData_kind(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":                      "test-disk",
-		"type":                      "pd-ssd",
-		"zone":                      "us-central1-a",
-		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name": "test-disk",
+		"type": "pd-ssd",
+		"zone": "us-central1-a",
+		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -42,10 +42,10 @@ func TestFakeResourceData_id(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":                      "test-disk",
-		"type":                      "pd-ssd",
-		"zone":                      "us-central1-a",
-		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name": "test-disk",
+		"type": "pd-ssd",
+		"zone": "us-central1-a",
+		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -60,10 +60,10 @@ func TestFakeResourceData_get(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":                      "test-disk",
-		"type":                      "pd-ssd",
-		"zone":                      "us-central1-a",
-		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name": "test-disk",
+		"type": "pd-ssd",
+		"zone": "us-central1-a",
+		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -78,10 +78,10 @@ func TestFakeResourceData_getOkOk(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":                      "test-disk",
-		"type":                      "pd-ssd",
-		"zone":                      "us-central1-a",
-		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name": "test-disk",
+		"type": "pd-ssd",
+		"zone": "us-central1-a",
+		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -98,10 +98,10 @@ func TestFakeResourceData_getOkNonexistentField(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":                      "test-disk",
-		"type":                      "pd-ssd",
-		"zone":                      "us-central1-a",
-		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name": "test-disk",
+		"type": "pd-ssd",
+		"zone": "us-central1-a",
+		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(

--- a/converters/google/fake_resource_data_test.go
+++ b/converters/google/fake_resource_data_test.go
@@ -24,10 +24,10 @@ func TestFakeResourceData_kind(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":  "test-disk",
-		"type":  "pd-ssd",
-		"zone":  "us-central1-a",
-		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name":                      "test-disk",
+		"type":                      "pd-ssd",
+		"zone":                      "us-central1-a",
+		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -35,18 +35,17 @@ func TestFakeResourceData_kind(t *testing.T) {
 		p.ResourcesMap["google_compute_disk"].Schema,
 		values,
 	)
-	assert.Equal(t, d.Kind(), "google_compute_disk")
+	assert.Equal(t, "google_compute_disk", d.Kind())
 }
-
 
 func TestFakeResourceData_id(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":  "test-disk",
-		"type":  "pd-ssd",
-		"zone":  "us-central1-a",
-		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name":                      "test-disk",
+		"type":                      "pd-ssd",
+		"zone":                      "us-central1-a",
+		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -57,15 +56,14 @@ func TestFakeResourceData_id(t *testing.T) {
 	assert.Equal(t, d.Id(), "")
 }
 
-
 func TestFakeResourceData_get(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":  "test-disk",
-		"type":  "pd-ssd",
-		"zone":  "us-central1-a",
-		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name":                      "test-disk",
+		"type":                      "pd-ssd",
+		"zone":                      "us-central1-a",
+		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -76,15 +74,14 @@ func TestFakeResourceData_get(t *testing.T) {
 	assert.Equal(t, d.Get("name"), "test-disk")
 }
 
-
 func TestFakeResourceData_getOkOk(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":  "test-disk",
-		"type":  "pd-ssd",
-		"zone":  "us-central1-a",
-		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name":                      "test-disk",
+		"type":                      "pd-ssd",
+		"zone":                      "us-central1-a",
+		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -93,19 +90,18 @@ func TestFakeResourceData_getOkOk(t *testing.T) {
 		values,
 	)
 	res, ok := d.GetOk("name")
-	assert.Equal(t, res, "test-disk")
+	assert.Equal(t, "test-disk", res)
 	assert.True(t, ok)
 }
 
-
-func TestFakeResourceData_getOkNotOk(t *testing.T) {
+func TestFakeResourceData_getOkNonexistentField(t *testing.T) {
 	p := provider.Provider()
 
 	values := map[string]interface{}{
-		"name":  "test-disk",
-		"type":  "pd-ssd",
-		"zone":  "us-central1-a",
-		"image": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+		"name":                      "test-disk",
+		"type":                      "pd-ssd",
+		"zone":                      "us-central1-a",
+		"image":                     "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
 		"physical_block_size_bytes": 4096,
 	}
 	d := NewFakeResourceData(
@@ -115,5 +111,216 @@ func TestFakeResourceData_getOkNotOk(t *testing.T) {
 	)
 	res, ok := d.GetOk("incorrect")
 	assert.Nil(t, res)
+	assert.False(t, ok)
+}
+
+func TestFakeResourceData_getOkEmptyString(t *testing.T) {
+	p := provider.Provider()
+
+	values := map[string]interface{}{
+		"name":                      "test-disk",
+		"type":                      "pd-ssd",
+		"zone":                      "us-central1-a",
+		"image":                     "",
+		"physical_block_size_bytes": 4096,
+	}
+	d := NewFakeResourceData(
+		"google_compute_disk",
+		p.ResourcesMap["google_compute_disk"].Schema,
+		values,
+	)
+	res, ok := d.GetOk("image")
+	assert.Equal(t, "", res)
+	assert.False(t, ok)
+}
+
+func TestFakeResourceData_getOkUnsetString(t *testing.T) {
+	p := provider.Provider()
+
+	values := map[string]interface{}{
+		"name":     "my-node-pool",
+		"location": "us-central1",
+		"cluster":  "projects/my-project-id/global/clusters/my-gke-cluster",
+		"config": map[string]interface{}{
+			"machineType": "n1-standard-1",
+			"metadata": map[string]string{
+				"disable-legacy-endpoints": "true",
+			},
+			"oauthScopes": []string{
+				"https://www.googleapis.com/auth/cloud-platform",
+			},
+			"preemptible": true,
+		},
+	}
+	d := NewFakeResourceData(
+		"google_container_cluster",
+		p.ResourcesMap["google_container_cluster"].Schema,
+		values,
+	)
+	res, ok := d.GetOk("subnetwork")
+	assert.Equal(t, "", res)
+	assert.False(t, ok)
+}
+
+func TestFakeResourceData_getOkTypeObject(t *testing.T) {
+	p := provider.Provider()
+
+	values := map[string]interface{}{
+		"advanced_machine_features": []interface{}{},
+		"allow_stopping_for_update": nil,
+		"attached_disk": []interface{}{
+			map[string]interface{}{
+				"device_name":             "test-device_name",
+				"disk_encryption_key_raw": nil,
+				"kms_key_self_link":       "test-kms_key_self_link",
+				"mode":                    "READ_ONLY",
+				"source":                  "test-source",
+			},
+			map[string]interface{}{
+				"disk_encryption_key_raw": nil,
+				"mode":                    "READ_WRITE",
+				"source":                  "test-source2",
+			},
+		},
+		"boot_disk": []interface{}{
+			map[string]interface{}{
+				"auto_delete":             true,
+				"disk_encryption_key_raw": nil,
+				"initialize_params": []interface{}{
+					map[string]interface{}{
+						"image": "debian-cloud/debian-9",
+					},
+				},
+				"mode": "READ_WRITE",
+			},
+		},
+		"can_ip_forward":          false,
+		"deletion_protection":     false,
+		"description":             nil,
+		"desired_status":          nil,
+		"enable_display":          nil,
+		"hostname":                nil,
+		"labels":                  nil,
+		"machine_type":            "n1-standard-1",
+		"metadata":                nil,
+		"metadata_startup_script": nil,
+		"name":                    "test",
+		"network_interface": []interface{}{
+			map[string]interface{}{
+				"access_config": []interface{}{
+					map[string]interface{}{
+						"public_ptr_domain_name": nil,
+					},
+				},
+				"alias_ip_range":     []interface{}{},
+				"ipv6_access_config": []interface{}{},
+				"network":            "default",
+				"nic_type":           nil,
+			},
+		},
+		"resource_policies": nil,
+		"scratch_disk": []interface{}{
+			map[string]interface{}{
+				"interface": "SCSI",
+			},
+		},
+		"service_account":          []interface{}{},
+		"shielded_instance_config": []interface{}{},
+		"tags": []interface{}{
+			"bar",
+			"foo",
+		},
+		"timeouts": nil,
+		"zone":     "us-central1-a",
+	}
+	d := NewFakeResourceData(
+		"google_compute_instance",
+		p.ResourcesMap["google_compute_instance"].Schema,
+		values,
+	)
+	res, ok := d.GetOk("attached_disk.0")
+	assert.Equal(t, map[string]interface{}{
+		"device_name":                "test-device_name",
+		"disk_encryption_key_raw":    "",
+		"disk_encryption_key_sha256": "",
+		"kms_key_self_link":          "test-kms_key_self_link",
+		"mode":                       "READ_ONLY",
+		"source":                     "test-source",
+	}, res)
+	assert.True(t, ok)
+}
+
+func TestFakeResourceData_getOknsetTypeObject(t *testing.T) {
+	p := provider.Provider()
+
+	values := map[string]interface{}{
+		"advanced_machine_features": []interface{}{},
+		"allow_stopping_for_update": nil,
+		"attached_disk":             []interface{}{},
+		"boot_disk": []interface{}{
+			map[string]interface{}{
+				"auto_delete":             true,
+				"disk_encryption_key_raw": nil,
+				"initialize_params": []interface{}{
+					map[string]interface{}{
+						"image": "debian-cloud/debian-9",
+					},
+				},
+				"mode": "READ_WRITE",
+			},
+		},
+		"can_ip_forward":          false,
+		"deletion_protection":     false,
+		"description":             nil,
+		"desired_status":          nil,
+		"enable_display":          nil,
+		"hostname":                nil,
+		"labels":                  nil,
+		"machine_type":            "n1-standard-1",
+		"metadata":                nil,
+		"metadata_startup_script": nil,
+		"name":                    "test",
+		"network_interface": []interface{}{
+			map[string]interface{}{
+				"access_config": []interface{}{
+					map[string]interface{}{
+						"public_ptr_domain_name": nil,
+					},
+				},
+				"alias_ip_range":     []interface{}{},
+				"ipv6_access_config": []interface{}{},
+				"network":            "default",
+				"nic_type":           nil,
+			},
+		},
+		"resource_policies": nil,
+		"scratch_disk": []interface{}{
+			map[string]interface{}{
+				"interface": "SCSI",
+			},
+		},
+		"service_account":          []interface{}{},
+		"shielded_instance_config": []interface{}{},
+		"tags": []interface{}{
+			"bar",
+			"foo",
+		},
+		"timeouts": nil,
+		"zone":     "us-central1-a",
+	}
+	d := NewFakeResourceData(
+		"google_compute_instance",
+		p.ResourcesMap["google_compute_instance"].Schema,
+		values,
+	)
+	res, ok := d.GetOk("attached_disk.0")
+	assert.Equal(t, map[string]interface{}{
+		"device_name":                "",
+		"disk_encryption_key_raw":    "",
+		"disk_encryption_key_sha256": "",
+		"kms_key_self_link":          "",
+		"mode":                       "",
+		"source":                     "",
+	}, res)
 	assert.False(t, ok)
 }

--- a/test/cli_test.go
+++ b/test/cli_test.go
@@ -63,6 +63,7 @@ func TestCLI(t *testing.T) {
 		{name: "example_bigquery_dataset"},
 		{name: "example_bigtable_instance"},
 		{name: "example_compute_disk"},
+		{name: "example_compute_disk_empty_image"},
 		{name: "example_compute_firewall"},
 		{name: "example_compute_forwarding_rule"},
 		{name: "example_compute_instance"},

--- a/test/read_test.go
+++ b/test/read_test.go
@@ -105,8 +105,8 @@ func TestReadPlannedAssetsCoverage(t *testing.T) {
 				t.Fatalf("ReadPlannedAssets(%s, %s, %s, %t): %v", planfile, data.Provider["project"], data.Ancestry, true, err)
 			}
 
-			expectedAssets := normalizeAssets(t, got, true)
-			actualAssets := normalizeAssets(t, want, true)
+			expectedAssets := normalizeAssets(t, want, true)
+			actualAssets := normalizeAssets(t, got, true)
 			require.ElementsMatch(t, actualAssets, expectedAssets)
 		})
 	}

--- a/test/read_test.go
+++ b/test/read_test.go
@@ -24,6 +24,7 @@ func TestReadPlannedAssetsCoverage(t *testing.T) {
 		{name: "example_bigquery_dataset"},
 		{name: "example_bigtable_instance"},
 		{name: "example_compute_disk"},
+		{name: "example_compute_disk_empty_image"},
 		{name: "example_compute_firewall"},
 		// This test can't run in offline mode
 		// {name: "example_compute_instance"},

--- a/testdata/templates/bucket.json
+++ b/testdata/templates/bucket.json
@@ -9,8 +9,9 @@
       "discovery_name": "Bucket",
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
-        "billing": {},
-        "lifecycle": {},
+        "lifecycle": {
+          "rule": []
+        },
         "iamConfiguration": {
           "uniformBucketLevelAccess": {
             "enabled": false

--- a/testdata/templates/example_compute_disk_empty_image.json
+++ b/testdata/templates/example_compute_disk_empty_image.json
@@ -14,7 +14,6 @@
         },
         "name": "test-disk",
         "physicalBlockSizeBytes": 4096,
-        "sourceImage": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
         "type": "projects/{{.Provider.project}}/zones/us-central1-a/diskTypes/pd-ssd",
         "zone": "projects/{{.Provider.project}}/global/zones/us-central1-a"
       }

--- a/testdata/templates/example_compute_disk_empty_image.json
+++ b/testdata/templates/example_compute_disk_empty_image.json
@@ -1,0 +1,23 @@
+[
+  {
+    "name": "//compute.googleapis.com/projects/{{.Provider.project}}/zones/us-central1-a/disks/test-disk",
+    "asset_type": "compute.googleapis.com/Disk",
+    "ancestry_path": "{{.Ancestry}}/project/{{.Provider.project}}",
+    "resource": {
+      "version": "v1",
+      "discovery_document_uri": "https://www.googleapis.com/discovery/v1/apis/compute/v1/rest",
+      "discovery_name": "Disk",
+      "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
+      "data": {
+        "labels": {
+          "environment": "dev"
+        },
+        "name": "test-disk",
+        "physicalBlockSizeBytes": 4096,
+        "sourceImage": "projects/debian-cloud/global/images/debian-8-jessie-v20170523",
+        "type": "projects/{{.Provider.project}}/zones/us-central1-a/diskTypes/pd-ssd",
+        "zone": "projects/{{.Provider.project}}/global/zones/us-central1-a"
+      }
+    }
+  }
+]

--- a/testdata/templates/example_compute_disk_empty_image.tf
+++ b/testdata/templates/example_compute_disk_empty_image.tf
@@ -1,0 +1,39 @@
+/**
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+terraform {
+  required_providers {
+    google = {
+      source = "hashicorp/google"
+      version = "~> {{.Provider.version}}"
+    }
+  }
+}
+
+provider "google" {
+  {{if .Provider.credentials }}credentials = "{{.Provider.credentials}}"{{end}}
+}
+
+resource "google_compute_disk" "default" {
+  name  = "test-disk"
+  type  = "pd-ssd"
+  zone  = "us-central1-a"
+  image = ""
+  labels = {
+    environment = "dev"
+  }
+  physical_block_size_bytes = 4096
+}

--- a/testdata/templates/example_compute_disk_empty_image.tfplan.json
+++ b/testdata/templates/example_compute_disk_empty_image.tfplan.json
@@ -1,0 +1,131 @@
+{
+  "format_version": "0.1",
+  "terraform_version": "0.12.10",
+  "planned_values": {
+    "root_module": {
+      "resources": [
+        {
+          "address": "google_compute_disk.default",
+          "mode": "managed",
+          "type": "google_compute_disk",
+          "name": "default",
+          "provider_name": "google",
+          "schema_version": 0,
+          "values": {
+            "description": null,
+            "disk_encryption_key": [],
+            "disk_encryption_key_raw": null,
+            "image": "",
+            "labels": {
+              "environment": "dev"
+            },
+            "name": "test-disk",
+            "physical_block_size_bytes": 4096,
+            "snapshot": null,
+            "source_image_encryption_key": [],
+            "source_snapshot_encryption_key": [],
+            "timeouts": null,
+            "type": "pd-ssd",
+            "zone": "us-central1-a"
+          }
+        }
+      ]
+    }
+  },
+  "resource_changes": [
+    {
+      "address": "google_compute_disk.default",
+      "mode": "managed",
+      "type": "google_compute_disk",
+      "name": "default",
+      "provider_name": "google",
+      "change": {
+        "actions": [
+          "create"
+        ],
+        "before": null,
+        "after": {
+          "description": null,
+          "disk_encryption_key": [],
+          "disk_encryption_key_raw": null,
+          "image": "",
+          "labels": {
+            "environment": "dev"
+          },
+          "name": "test-disk",
+          "physical_block_size_bytes": 4096,
+          "snapshot": null,
+          "source_image_encryption_key": [],
+          "source_snapshot_encryption_key": [],
+          "timeouts": null,
+          "type": "pd-ssd",
+          "zone": "us-central1-a"
+        },
+        "after_unknown": {
+          "creation_timestamp": true,
+          "disk_encryption_key": [],
+          "disk_encryption_key_sha256": true,
+          "id": true,
+          "label_fingerprint": true,
+          "labels": {},
+          "last_attach_timestamp": true,
+          "last_detach_timestamp": true,
+          "project": true,
+          "self_link": true,
+          "size": true,
+          "source_image_encryption_key": [],
+          "source_image_id": true,
+          "source_snapshot_encryption_key": [],
+          "source_snapshot_id": true,
+          "users": true
+        }
+      }
+    }
+  ],
+  "configuration": {
+    "provider_config": {
+      "google": {
+        "name": "google",
+        "expressions": {
+          "project": {
+            "constant_value": "{{.Provider.project}}"
+          }
+        }
+      }
+    },
+    "root_module": {
+      "resources": [
+        {
+          "address": "google_compute_disk.default",
+          "mode": "managed",
+          "type": "google_compute_disk",
+          "name": "default",
+          "provider_config_key": "google",
+          "expressions": {
+            "image": {
+              "constant_value": ""
+            },
+            "labels": {
+              "constant_value": {
+                "environment": "dev"
+              }
+            },
+            "name": {
+              "constant_value": "test-disk"
+            },
+            "physical_block_size_bytes": {
+              "constant_value": 4096
+            },
+            "type": {
+              "constant_value": "pd-ssd"
+            },
+            "zone": {
+              "constant_value": "us-central1-a"
+            }
+          },
+          "schema_version": 0
+        }
+      ]
+    }
+  }
+}

--- a/testdata/templates/example_compute_instance.json
+++ b/testdata/templates/example_compute_instance.json
@@ -10,14 +10,6 @@
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
         "canIpForward": false,
-        "displayDevice": {
-          "enableDisplay": false
-        },
-        "shieldedInstanceConfig": {
-          "enableIntegrityMonitoring": false,
-          "enableSecureBoot": false,
-          "enableVtpm": false
-        },
         "deletionProtection": false,
         "disks": [
           {

--- a/testdata/templates/example_project_in_folder.json
+++ b/testdata/templates/example_project_in_folder.json
@@ -1,21 +1,5 @@
 [
     {
-      "name": "//cloudbilling.googleapis.com/projects/foobat/billingInfo",
-      "asset_type": "cloudbilling.googleapis.com/ProjectBillingInfo",
-      "ancestry_path": "organization/unknown/folder/{{.FolderID}}/project/foobat",
-      "resource": {
-        "version": "v1",
-        "discovery_document_uri": "https://www.googleapis.com/discovery/v1/apis/cloudbilling/v1/rest",
-        "discovery_name": "ProjectBillingInfo",
-        "parent": "//cloudresourcemanager.googleapis.com/projects/foobat",
-        "data": {
-          "billingAccountName": "billingAccounts/",
-          "name": "projects/foobat/billingInfo",
-          "projectId": "foobat"
-        }
-      }
-    },
-    {
       "name": "//cloudresourcemanager.googleapis.com/projects/foobat",
       "asset_type": "cloudresourcemanager.googleapis.com/Project",
       "ancestry_path": "organization/unknown/folder/{{.FolderID}}/project/foobat",

--- a/testdata/templates/example_project_in_org.json
+++ b/testdata/templates/example_project_in_org.json
@@ -1,21 +1,5 @@
 [
     {
-      "name": "//cloudbilling.googleapis.com/projects/foobat/billingInfo",
-      "asset_type": "cloudbilling.googleapis.com/ProjectBillingInfo",
-      "ancestry_path": "organization/{{.OrgID}}/project/foobat",
-      "resource": {
-        "version": "v1",
-        "discovery_document_uri": "https://www.googleapis.com/discovery/v1/apis/cloudbilling/v1/rest",
-        "discovery_name": "ProjectBillingInfo",
-        "parent": "//cloudresourcemanager.googleapis.com/projects/foobat",
-        "data": {
-          "billingAccountName": "billingAccounts/",
-          "name": "projects/foobat/billingInfo",
-          "projectId": "foobat"
-        }
-      }
-    },
-    {
       "name": "//cloudresourcemanager.googleapis.com/projects/foobat",
       "asset_type": "cloudresourcemanager.googleapis.com/Project",
       "ancestry_path": "organization/{{.OrgID}}/project/foobat",

--- a/testdata/templates/example_storage_bucket.json
+++ b/testdata/templates/example_storage_bucket.json
@@ -9,13 +9,14 @@
       "discovery_name": "Bucket",
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
-        "billing": {},
         "iamConfiguration": {
           "uniformBucketLevelAccess": {
             "enabled": false
           }
         },
-        "lifecycle": {},
+        "lifecycle": {
+          "rule": []
+        },
         "location": "EU",
         "name": "image-store-bucket",
         "project": "{{.Provider.project}}",

--- a/testdata/templates/example_storage_bucket_iam_binding.json
+++ b/testdata/templates/example_storage_bucket_iam_binding.json
@@ -9,13 +9,14 @@
       "discovery_name": "Bucket",
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
-        "billing": {},
         "iamConfiguration": {
           "uniformBucketLevelAccess": {
             "enabled": true
           }
         },
-        "lifecycle": {},
+        "lifecycle": {
+          "rule": []
+        },
         "location": "EU",
         "name": "fake-bucket-123456",
         "project": "{{.Provider.project}}",

--- a/testdata/templates/example_storage_bucket_iam_member.json
+++ b/testdata/templates/example_storage_bucket_iam_member.json
@@ -9,13 +9,14 @@
       "discovery_name": "Bucket",
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
-        "billing": {},
         "iamConfiguration": {
           "uniformBucketLevelAccess": {
             "enabled": true
           }
         },
-        "lifecycle": {},
+        "lifecycle": {
+          "rule": []
+        },
         "location": "EU",
         "name": "fake-bucket-123456",
         "project": "{{.Provider.project}}",

--- a/testdata/templates/example_storage_bucket_iam_member_random_suffix.json
+++ b/testdata/templates/example_storage_bucket_iam_member_random_suffix.json
@@ -9,13 +9,14 @@
       "discovery_name": "Bucket",
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
-        "billing": {},
         "iamConfiguration": {
           "uniformBucketLevelAccess": {
             "enabled": true
           }
         },
-        "lifecycle": {},
+        "lifecycle": {
+          "rule": []
+        },
         "location": "EU",
         "project": "{{.Provider.project}}",
         "storageClass": "STANDARD"

--- a/testdata/templates/example_storage_bucket_iam_policy.json
+++ b/testdata/templates/example_storage_bucket_iam_policy.json
@@ -9,13 +9,14 @@
       "discovery_name": "Bucket",
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
-        "billing": {},
         "iamConfiguration": {
           "uniformBucketLevelAccess": {
             "enabled": true
           }
         },
-        "lifecycle": {},
+        "lifecycle": {
+          "rule": []
+        },
         "location": "EU",
         "name": "fake-bucket-123456",
         "project": "{{.Provider.project}}",

--- a/testdata/templates/full_compute_instance.json
+++ b/testdata/templates/full_compute_instance.json
@@ -10,9 +10,6 @@
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
         "canIpForward": true,
-        "displayDevice": {
-          "enableDisplay": false
-        },
         "shieldedInstanceConfig": {
           "enableIntegrityMonitoring": true,
           "enableSecureBoot": true,
@@ -153,14 +150,6 @@
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
         "canIpForward": false,
-        "displayDevice": {
-          "enableDisplay": false
-        },
-        "shieldedInstanceConfig": {
-          "enableIntegrityMonitoring": false,
-          "enableSecureBoot": false,
-          "enableVtpm": false
-        },
         "deletionProtection": false,
         "disks": [
           {

--- a/testdata/templates/instance.json
+++ b/testdata/templates/instance.json
@@ -10,14 +10,6 @@
       "parent": "//cloudresourcemanager.googleapis.com/projects/{{.Provider.project}}",
       "data": {
         "canIpForward": false,
-        "displayDevice": {
-          "enableDisplay": false
-        },
-        "shieldedInstanceConfig": {
-          "enableIntegrityMonitoring": false,
-          "enableSecureBoot": false,
-          "enableVtpm": false
-        },
         "deletionProtection": false,
         "disks": [
           {

